blockchain/validate.go

```diff
@@ -774,14 +774,34 @@ func (b *BlockChain) checkBlockContext(block *btcutil.Block, prevNode *blockNode
 
 	fastAdd := flags&BFFastAdd == BFFastAdd
 	if !fastAdd {
+		medianTime := b.BestSnapshot().MedianTime
+
+		// Obtain the latest state of the deployed CSV soft-fork in
+		// order to properly guard the new validation behavior based on
+		// the current BIP 9 version bits state.
+		csvState, err := b.ThresholdState(chaincfg.DeploymentCSV)
```

Comment:
I've carried out your suggestion and have created a new internal version of  `ThresholdState` which assumes that the caller already has the `chainLock` acquired. This function is now used in this fragment in order to prevent the deadlock condition you pointed out.